### ISSUE TITLE
deferred tasks get kill during heartbeat callback in some rare cases

### ISSUE THREAD

**Issue by 0x26res** [apache/airflow#40435]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.8.1 (MWAA)

### What happened?

My airflow deferred (AWS) BatchOperator occasionally fail. When this happens I can see this:

```
{{local_task_job_runner.py:302}} WARNING - State of this instance has been externally set to deferred. Terminating instance.
```

I actually think it's an oversight in the code for local_task_job_runner. This [line](https://github.com/apache/airflow/blob/main/airflow/jobs/local_task_job_runner.py#L264) should check for state being RUNNING or DEFERRED. 

### How to reproduce

This is hard to reproduce. The issue is very transient.

### Operating System

MWAA

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.16.0

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

MWAA

### Anything else?

 I would say it happens one out of 100 run.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by wolfier** [comment]
I don't think a task in the deferred state should be counted as running. Once a task is deferred, it should exit as tasks do for other terminal states like success and failed. For the [LocalTaskJobRunner](https://github.com/apache/airflow/blob/2.8.1/airflow/jobs/local_task_job_runner.py#L284-L305) to produce the message `external set to deferred`, the task instance must have been in the deferred state for at least [scheduler_zombie_task_threshold](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#scheduler-zombie-task-threshold) seconds (see [source](https://github.com/apache/airflow/blob/2.8.1/airflow/jobs/local_task_job_runner.py#L167-L194)) and the corresponding StandardTaskRunner did not produced a return code during that time. The source code provides a helpful note that I think is what happened.
> \# potential race condition, the _run_raw_task commits `success` or other state
            # but task_runner does not exit right away due to slow process shutdown or any other reasons

The reporter also mentioned in [another comment](https://github.com/apache/airflow/pull/40453#issuecomment-2210569574) that their scheduler may not well resourced. However, this statement cannot be confirmed without the actual resource utilization graph.

> Also, I have a question related to the underlying issue. I've noticed this happens if a lot of deferred tasks are submitted at once. The server I'm running this on is very small (2 CPU, little memory). I think this is means the timing of the heart beat is messed up by a straved CPU.

Regardless, it would beneficial to see what happened to the task after being marked as deferred. The task log should tell us the exact timestamp. Compare that with when the LocalTaskJobRunner terminated the task, we can at least have a clearer view of the task/process lifecycle. I also think the full task log would help the investigation as the terminal code may provide additional context.

I suspect the issue is something to do with the StandardTaskRunner failing to exit for some reason. Considering deferred as a "running" state would probably let the LocalTaskJobRunner and StandardTaskRunner run indefinitely.

**Comment by 0x26res** [comment]
@wolfier thanks for the detailed answer. I've put the logs, I hopes it helps. 

For context our deferred task submit jobs to AWS batch and wait for completion. I don't suspect the submission would block for more than 300 seconds / [scheduler_zombie_task_threshold](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#scheduler-zombie-task-threshold)

But I know for fact this issue happens when a lot of these deferred tasks start at the same time.

**Comment by wolfier** [comment]
I'll start with the successful task / trigger execution. The task exited right away with the return code. The trigger fired as expected after the conditions are met.

```
[2024-08-03, 10:35:20 UTC] {{standard_task_runner.py:60}} INFO - Started process 3875 to run task
...
[2024-08-03, 10:35:21 UTC] {{taskinstance.py:2344}} INFO - Pausing task as DEFERRED. dag_id=eod_jobs, task_id=task_123, execution_date=20240802T001500, start_date=20240803T103519
[2024-08-03, 10:35:21 UTC] {{local_task_job_runner.py:231}} INFO - Task exited with return code 100 (task deferral)
...
[2024-08-03, 10:37:23 UTC] {{triggerer_job_runner.py:602}} INFO - Trigger eod_jobs/scheduled__2024-08-02T00:15:00+00:00/task_123/-1/2 (ID 5992) fired: TriggerEvent<{'status': 'success', 'job_id': 'JOB_UUID'}>
```

The failed task instance did not exit therefore does not have a return code. Given the logs, the reported behaviour is then expected.

```
[2024-08-03, 00:46:56 UTC] {{standard_task_runner.py:60}} INFO - Started process 26709 to run task
...
[2024-08-03, 00:47:01 UTC] {{taskinstance.py:2344}} INFO - Pausing task as DEFERRED. dag_id=eod_jobs, task_id=task_123, execution_date=20240802T001500, start_date=20240803T004656
...
[2024-08-03, 00:47:23 UTC] {{process_utils.py:131}} INFO - Sending 15 to group 26709. PIDs of all processes in the group: [26709]
[2024-08-03, 00:47:24 UTC] {{process_utils.py:86}} INFO - Sending the signal 15 to group 26709
[2024-08-03, 00:47:44 UTC] {{process_utils.py:79}} INFO - Process psutil.Process(pid=26709, status='terminated', exitcode=100, started='00:46:56') (26709) terminated with exit code 100
```

Surprisingly, the exit code was 100. 

We know that two consecutive calls of heartbeat_callback saw the `task_runner` return code is `None` and the task instance state was in `deferred` starting at `00:47:01`. We do not know what happened for 22 seconds.

I suspect one of the following happened:
* The execute function did complete and `return_code` is set to 100. However, the process has not exited. It may have gotten stuck at `os._exit(return_code)`. When the process was forced to exit via reaping which calls `os.killpg(process_group_id, sig)` , the `returncode` for the process was 100.
* The execute function did not complete. Not sure where it could be stuck.

**Comment by 0x26res** [comment]
@wolfier does each deferred task start a new process? It wouldn't surprised me, if there are a lot of deferred tasks starting/running at the same time that the overhead of each process causes some of them to be killed because OMM / lack of memory.

Does the scheduler throttle / limit the number of deferred tasks that are running at the same time?

**Comment by wolfier** [comment]
> does each deferred task start a new process? 

That's correct. Each task execution will spawn a new process. A deferrable operator task will be executed twice. Once to submit the trigger to the triggerer and another time to process the trigger event.

The process should be fairly short running though as you can see with the successful attempt. 

I am not sure how MWAA is setup but the triggerer and worker should be running in two different execution spaces. This means the number of concurrent triggers should not affect the worker.

> Does the scheduler throttle / limit the number of deferred tasks that are running at the same time?

The scheduler does not but the triggerer does have a [limit of triggers per triggerer](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#default-capacity).
